src/json2typescript/json-convert.ts

```diff
@@ -695,16 +695,18 @@ export class JsonConvert {
         // Check if mapping is defined
         if (typeof (mappings) === "undefined") return null;
 
-        // Get direct mapping if possible
-        const directMappingName: string = instance[Settings.CLASS_IDENTIFIER] + "." + propertyName;
-        if (typeof (mappings[directMappingName]) !== "undefined") {
-            return mappings[directMappingName];
-        }
-
-        // No mapping was found, try to find some
-        const indirectMappingNames: string[] = Object.keys(mappings).filter(key => key.match("\\." + propertyName + "$")); // use endsWidth in later versions
-        if (indirectMappingNames.length > 0) {
-            return mappings[indirectMappingNames[0]];
+        /* Find mapping by iterating up the prototype chain to find a matching mapping, rather than
+         * just searching by property name. */
+        let prototype = Object.getPrototypeOf(instance);
+        while (prototype != null) {
```

Comment:
Are you sure that your loop always terminates? Is there no such thing as a reference to self?

Following the docs, it should be fine:

> The prototype of the given object. If there are no inherited properties, null is returned.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/getPrototypeOf

Reply:
As long as the object is constructed normally, it should be fine.  You're right though, it's of course possible that someone has manually set the prototype to some self-referential or circular link.  I'll modify it to check for self as well.

Reply:
Actually, something I've just learned is that you can't in fact create circular prototype references.  That is, you can, but it will cause an error: "TypeError: Cyclic __proto__ value".  So we don't need to check for self - the chain will eventually always terminate.

Reply:
That's good to know! Otherwise this would be a quite frustrating user experience ;-)